dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,194 @@
+        var result = new Vector
+        {
+            Id = (string)keyObject,
+            Values = values.ToArray(),
+            Metadata = metadata,
+            SparseValues = null
+        };
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public TRecord MapFromStorageToDataModel(Vector storageModel, StorageToDataModelMapperOptions options)
+    {
+        var keyPropertyName = this._storagePropertyNames[this._keyPropertyInfo.Name];
+
+        var outputJsonObject = new JsonObject
+        {
+            { this._keyPropertyInfo.Name, JsonValue.Create(storageModel.Id) },
+        };
+
+        if (options?.IncludeVectors is true)
+        {
+            var propertyName = this._storagePropertyNames[this._vectorPropertyInfo.Name];
+            outputJsonObject.Add(this._vectorPropertyInfo.Name, new JsonArray(storageModel.Values.Select(x => JsonValue.Create(x)).ToArray()));
+        }
+
+        if (storageModel.Metadata != null)
+        {
+            foreach (var payloadProperty in this._payloadPropertiesInfo)
+            {
+                var propertyName = this._storagePropertyNames[payloadProperty.Name];
+                if (storageModel.Metadata.TryGetValue(propertyName, out var value))
+                {
+                    outputJsonObject.Add(payloadProperty.Name, ConvertFromMetadataValueToJsonNode(value));
+                }
+            }
+        }
+
+        return outputJsonObject.Deserialize<TRecord>()!;
+    }
+
+    private static JsonNode? ConvertFromMetadataValueToJsonNode(MetadataValue metadataValue)
+        => metadataValue.Inner switch
+        {
+            null => null,
+            bool boolValue => JsonValue.Create(boolValue),
+            string stringValue => JsonValue.Create(stringValue),
+            int intValue => JsonValue.Create(intValue),
+            long longValue => JsonValue.Create(longValue),
+            float floatValue => JsonValue.Create(floatValue),
+            double doubleValue => JsonValue.Create(doubleValue),
+            decimal decimalValue => JsonValue.Create(decimalValue),
+            MetadataValue[] array => new JsonArray(array.Select(ConvertFromMetadataValueToJsonNode).ToArray()),
+            _ => throw new VectorStoreRecordMappingException($"Unsupported metadata type: '{metadataValue.Inner?.GetType().FullName}'."),
+        };
+
+    // TODO: take advantage of MetadataValue.TryCreate once we upgrade the version of Pinecone.NET
+    private static MetadataValue ConvertToMetadataValue(object? sourceValue)
+        => sourceValue switch
+        {
+            bool boolValue => boolValue,
+            string stringValue => stringValue,
+            int intValue => intValue,
+            long longValue => longValue,
+            float floatValue => floatValue,
+            double doubleValue => doubleValue,
+            decimal decimalValue => decimalValue,
+            string[] stringArray => stringArray,
+            List<string> stringList => stringList,
```

Comment:
Would it make sense to support all string enumerables?  E.g. to check if source value `is IEnumerable<string>`, in case someone used one of the myriad collection types.  This would of course require changes where we verify the supported property types as well.
I added an optional param to `VerifyPropertyTypes`, `supportEnumerable` which checks for all these cases for any of the supplied types, but if Pinecone only supports string lists, instead of lists of any of the other types, that option will be to broad for you.  The bit of code that's there should work though.
Something to consider anyway.

Reply:
sounds good, although Pinecone.NET MetadataValue only supports lists and arrays, so for all other types we are calling ToArray(), perf and memory hit, but better flexibility.